integrated_channels/blackboard/client.py

```diff
@@ -40,15 +45,235 @@ def __init__(self, enterprise_configuration):
+
+        for course in course_response:
+            if course.get('externalId') == external_id:
+                course_id = course.get('id')
+                return course_id
+        return None
+
+    def _create_session(self):
+        """
+        Will only create a new session if token expiry has been reached
+        """
+        self.session, self.expires_at = refresh_session_if_expired(
+            self._get_oauth_access_token,
+            self.session,
+            self.expires_at,
+        )
+
+    def _get_oauth_access_token(self):
+        """Fetch access token using refresh_token workflow from Blackboard
+
+        Returns:
+            access_token (str): the OAuth access token to access the Blackboard server
+            expires_in (int): the number of seconds after which token will expire
+        Raises:
+            HTTPError: If we received a failure response code.
+            ClientError: If an unexpected response format was received that we could not parse.
+        """
+
+        if not self.enterprise_configuration.refresh_token:
+            raise ClientError(
+                "Failed to generate oauth access token: Refresh token required.",
+                HTTPStatus.INTERNAL_SERVER_ERROR.value
+            )
+
+        if (not self.enterprise_configuration.blackboard_base_url
+                or not self.config.oauth_token_auth_path):
+            raise ClientError(
+                "Failed to generate oauth access token: oauth path missing from configuration.",
+                HTTPStatus.INTERNAL_SERVER_ERROR.value
+            )
+        auth_token_url = urljoin(
+            self.enterprise_configuration.blackboard_base_url,
+            self.config.oauth_token_auth_path,
+        )
+
+        auth_token_params = {
+            'grant_type': 'refresh_token',
+            'refresh_token': self.enterprise_configuration.refresh_token,
+        }
+
+        auth_response = requests.post(
+            auth_token_url,
+            auth_token_params,
+            headers={
+                'Authorization': self._create_auth_header(),
+                'Content-Type': 'application/x-www-form-urlencoded'
+            }
+        )
+        if auth_response.status_code >= 400:
+            raise ClientError(auth_response.text, auth_response.status_code)
+        try:
+            data = auth_response.json()
+            # do not forget to save the new refresh token otherwise subsequent requests will fail
+            self.enterprise_configuration.refresh_token = data["refresh_token"]
+            self.enterprise_configuration.save()
+            return data['access_token'], data["expires_in"]
+        except (KeyError, ValueError):
+            raise ClientError(auth_response.text, auth_response.status_code)
+
+    def _create_auth_header(self):
+        """
+        auth header in oauth2 token format as required by blackboard doc
+        """
+        if not self.enterprise_configuration.client_id:
```

Comment:
I have mixed feelings about creating static methods to verify variable existence vs just checking inline, but I do feel like we should be consistent.

Reply:
How do you feel about either doing something like https://github.com/edx/edx-enterprise/pull/996/files#diff-e8224b0f5aaaee88283c12d5d533f5700f33f2e63c7939e93f814ef94bf95716R163? Or if not, then I don't see why we should have those static methods.

Reply:
yeah looks like that one is either way for me, it's a style thing. they are more accurately validator methods I think. stuff like does externalId exist? does course_id valid? and since they are used in more than once place I figured it would be better to DRY it up

I thikn if we called these validate_xyz instead of raise_for_ then it starts to make a bit more sense, what'cha think?

Reply:
the reason I think a method makes sense btw, is that it's very possible we want to validate more things about the payload in the future, so just thought would be a nice convenient place to put them (also makes it easier for us to tune the exception message in the future or do logging in one place, if we want to do those things for those cases)